File: solutions/zig/sieve/1/sieve.zig
const std = @import("std");

pub fn primes(buffer: []u32, comptime limit: u32) []u32 {
    var isPrime = std.StaticBitSet(limit + 1).initFull();
    if (limit < 2) return buffer[0..0];
    var numPrimes: usize = 0;
    for (2..limit + 1) |i| {
        if (isPrime.isSet(i)) {
            buffer[numPrimes] = @intCast(i);
            numPrimes += 1;
            var num = i * i;
            while (num <= limit) : (num += i) {
                isPrime.unset(num);
            }
        }
    }
    return buffer[0..numPrimes];
}

File: solutions/fortran/sieve/1/sieve.f90
module sieve
   implicit none

contains

   function primes(limit) result(array)
      integer, intent(in) :: limit
      integer, allocatable :: array(:)

      logical, allocatable :: isPrime(:)
      integer :: i, j, numPrimes

      allocate(isPrime(limit))
      isPrime(:) = .true.

      numPrimes = 0
      do i = 2, limit
         if (isPrime(i)) then
            do j = i*i, limit, i
               isPrime(j) = .false.
            end do
            numPrimes = numPrimes + 1
         end if
      end do

      allocate(array(numPrimes))

      i = 1
      do j = 2, limit
         if (isPrime(j))  then
            array(i) = j
            i = i + 1
         end if
      end do




   end function primes

end module sieve
